--- dbus-0.60/bus/messagebus.in.start-early	2005-01-18 17:20:38.000000000 -0500
+++ dbus-0.60/bus/messagebus.in	2006-01-20 18:03:33.000000000 -0500
@@ -2,7 +2,7 @@
 #
 # messagebus:   The D-BUS systemwide message bus
 #
-# chkconfig: 345 97 03
+# chkconfig: 345 22 85 
 # description:  This is a daemon which broadcasts notifications of system events \
 #               and other messages. See http://www.freedesktop.org/software/dbus/
 #
@@ -11,7 +11,7 @@
 #
 
 # Sanity checks.
-[ -x @EXPANDED_BINDIR@/dbus-daemon ] || exit 0
+[ -x /bin/dbus-daemon ] || exit 0
 
 # Source function library.
 . @EXPANDED_SYSCONFDIR@/rc.d/init.d/functions
