--- dbus-0.60/bus/messagebus.in.start-early	2006-01-20 16:39:46.000000000 -0500
+++ dbus-0.60/bus/messagebus.in	2006-01-20 16:42:06.000000000 -0500
@@ -2,7 +2,7 @@
 #
 # messagebus:   The D-BUS systemwide message bus
 #
-# chkconfig: 345 97 03
+# chkconfig: 345 22 85 
 # description:  This is a daemon which broadcasts notifications of system events \
 #               and other messages. See http://www.freedesktop.org/software/dbus/
 #
