--- dbus-1.0.2/bus/bus.c.audit_session	2006-12-11 14:21:23.000000000 -0500
+++ dbus-1.0.2/bus/bus.c	2007-03-30 11:53:46.000000000 -0400
@@ -746,11 +746,6 @@
       _dbus_string_free (&pid);
     }
 
-  if (!bus_selinux_full_init ())
-    {
-      _dbus_warn ("SELinux initialization failed\n");
-    }
-
   if (!process_config_postinit (context, parser, error))
     {
       _DBUS_ASSERT_ERROR_IS_SET (error);
@@ -775,6 +770,11 @@
 	}
     }
   
+  if (!bus_selinux_full_init ())
+    {
+      _dbus_warn ("SELinux initialization failed\n");
+    }
+
   dbus_server_free_data_slot (&server_data_slot);
   
   return context;
